### ISSUE TITLE
Disallow async constructors

### ISSUE THREAD

**Issue by benthecarman** [rustwasm/wasm-bindgen#3976]
### Describe the Bug

wasm_bindgen allows you to create an async constructor even though this is not supported by typescript.

### Steps to Reproduce

```rust

#[wasm_bindgen]
pub struct MyStruct {
	data: u32
}

#[wasm_bindgen]
impl MyStruct {
  #[wasm_bindgen(constructor)]
  pub async fn new(data: u32) -> Self {
     // do something async
     Self { data }
  }
}
```

This compiles and actually works, however seems like undefined behavior and should just fail to compile

### Expected Behavior

Unknown, probably fail to compile

### Actual Behavior

Works

**Comment by Liamolucko** [comment]
We considered removing support for this in #3562, since it was originally a bug, but discovered that people had started relying on it and so left it to avoid breakage.

You can see lots of instances of people doing this with https://github.com/search?q=%2F%23%5C%5Bwasm_bindgen%5C%28constructor%5C%29%5C%5D%5B+%5Cn%5D*pub+async+fn%2F&type=code.

**Comment by benthecarman** [comment]
Do you know what actually ends up happening under the hood, is it safe?

**Comment by Liamolucko** [comment]
> Do you know what actually ends up happening under the hood, is it safe?

Yes; `wasm-bindgen` treats constructors nearly the same way as regular associated functions, and so it generates code based on the return type of the constructor (which for async functions is a `JsValue` after desugaring) rather than the type it's attached to.

This is what the generated JS code ends up looking like:

```js
export class MyStruct {
    // ...
    constructor(data) {
        const ret = wasm.mystruct_new(data);
        return takeObject(ret);
    }
}
```

JS actually supports `return`ing from constructors, and if the returned value is something other than `this` then `new MyStruct()` just returns that value instead of actually making a `MyStruct`.

**Comment by daxpedda** [comment]
Maybe we should investigate if we can emit a warning in this case (with the help of `deprecated`).
